core/src/main/scala/org/apache/spark/sql/catalyst/expressions/ParamLiteral.scala

```diff
@@ -0,0 +1,168 @@
+          LiteralValue(value, pos))
+        val valueTerm = ctx.freshName("value")
+        ev.copy(
+          s"""
+             |final boolean $isNull = $valueRef.value() == null;
+             |final boolean $valueTerm = $isNull ? ${ctx.defaultValue(dataType)}
+             |    : ((Boolean)$valueRef.value()).booleanValue();
+           """.stripMargin, isNull, valueTerm)
+      case FloatType =>
+        val isNull = ctx.freshName("isNull")
+        assert(value.isInstanceOf[Float], s"unexpected type $dataType instead of FloatType")
+        val valueRef = ctx.addReferenceObj("literal",
+          LiteralValue(value, pos))
+        val valueTerm = ctx.freshName("value")
+        ev.copy(
+          s"""
+             |final boolean $isNull = $valueRef.value() == null;
+             |final float $valueTerm = $isNull ? ${ctx.defaultValue(dataType)}
+             |    : ((Float)$valueRef.value()).floatValue();
+           """.stripMargin, isNull, valueTerm)
+      case DoubleType =>
+        val isNull = ctx.freshName("isNull")
+        assert(value.isInstanceOf[Double], s"unexpected type $dataType instead of DoubleType")
+        val valueRef = ctx.addReferenceObj("literal",
+          LiteralValue(value, pos))
+        val valueTerm = ctx.freshName("value")
+        ev.copy(
+          s"""
+             |final boolean $isNull = $valueRef.value() == null;
+             |final double $valueTerm = $isNull ? ${ctx.defaultValue(dataType)}
+             |    : ((Double)$valueRef.value()).doubleValue();
+           """.stripMargin, isNull, valueTerm)
+      case ByteType =>
+        val isNull = ctx.freshName("isNull")
+        assert(value.isInstanceOf[Byte], s"unexpected type $dataType instead of ByteType")
+        val valueRef = ctx.addReferenceObj("literal",
+          LiteralValue(value, pos))
+        val valueTerm = ctx.freshName("value")
+        ev.copy(
+          s"""
+             |final boolean $isNull = $valueRef.value() == null;
+             |final byte $valueTerm = $isNull ? ${ctx.defaultValue(dataType)}
+             |    : ((Byte)$valueRef.value()).byteValue();
+           """.stripMargin, isNull, valueTerm)
+      case ShortType =>
+        val isNull = ctx.freshName("isNull")
+        assert(value.isInstanceOf[Short], s"unexpected type $dataType instead of ShortType")
+        val valueRef = ctx.addReferenceObj("literal",
+          LiteralValue(value, pos))
+        val valueTerm = ctx.freshName("value")
+        ev.copy(
+          s"""
+             |final boolean $isNull = $valueRef.value() == null;
+             |final short $valueTerm = $isNull ? ${ctx.defaultValue(dataType)}
+             |    : ((Short)$valueRef.value()).shortValue();
+           """.stripMargin, isNull, valueTerm)
+      case IntegerType | DateType =>
+        val isNull = ctx.freshName("isNull")
+        assert(value.isInstanceOf[Int], s"unexpected type $dataType instead of DateType or IntegerType")
```

Comment:
can be rewritten as
```scala
      case t@(IntegerType | DateType) =>
        val isNull = ctx.freshName("isNull")
        assert(value.isInstanceOf[Int], s"unexpected type $dataType instead of $t")
```